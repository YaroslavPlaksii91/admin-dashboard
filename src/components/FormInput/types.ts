import { UseFormRegisterReturn, FieldErrors } from 'react-hook-form';

export type FormInputProps = {
  label: string;
  type: string;
  name: string;
  placeholder: string;
  register: UseFormRegisterReturn;
  errors: FieldErrors;
  isPassword?: boolean;
  defaultValue?: string;
};
