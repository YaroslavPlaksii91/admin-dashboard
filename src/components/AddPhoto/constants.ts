export const IMAGE_SIZE = Object.freeze({
  SMALL: 44,
  BIG: 60,
});
