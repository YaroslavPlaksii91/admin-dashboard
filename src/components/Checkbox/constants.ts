export const CHECKBOX_TEST_ID = 'checkbox';
