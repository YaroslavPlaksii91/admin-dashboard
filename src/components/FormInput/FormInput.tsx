import { FC, useState, MouseEvent } from 'react';
import { TextField, InputAdornment, IconButton } from '@mui/material';
import { Visibility, VisibilityOff } from '@mui/icons-material';

import { FormInputProps } from './types';

export const FormInput: FC<FormInputProps> = ({
  label,
  type,
  name,
  placeholder,
  register,
  errors,
  isPassword,
}) => {
  const [showPassword, setShowPassword] = useState(false);

  const handleClickShowPassword = () => setShowPassword(show => !show);

  const handleMouseDownPassword = (event: MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
  };

  return (
    <TextField
      label={label}
      type={isPassword && showPassword ? 'text' : type}
      placeholder={placeholder}
      variant="outlined"
      fullWidth
      {...register}
      error={!!errors[name]}
      helperText={errors[name]?.message as string}
      InputLabelProps={{
        shrink: true,
      }}
      InputProps={
        isPassword
          ? {
              endAdornment: (
                <InputAdornment position="end">
                  <IconButton
                    aria-label="toggle password visibility"
                    onClick={handleClickShowPassword}
                    onMouseDown={handleMouseDownPassword}
                    edge="end"
                  >
                    {showPassword ? <VisibilityOff /> : <Visibility />}
                  </IconButton>
                </InputAdornment>
              ),
            }
          : undefined
      }
    />
  );
};
