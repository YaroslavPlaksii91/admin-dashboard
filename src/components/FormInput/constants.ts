export const TOGGLE_BUTTON_TEST_ID = 'toggleButton';
