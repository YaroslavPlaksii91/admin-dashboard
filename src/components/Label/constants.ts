export const LABEL_TYPES = Object.freeze({
  ROUND: 'round',
  SQUARE: 'square',
});

export const LABEL_COLORS = Object.freeze({
  DEFAULT: 'default',
  SUCCESS: 'success',
  WARNING: 'warning',
  ERROR: 'error',
});
