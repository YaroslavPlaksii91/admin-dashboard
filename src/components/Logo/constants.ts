export const LOGO_TYPES = Object.freeze({
  VERTICAL: 'vertical',
  HORIZONTAL: 'horizontal',
});
