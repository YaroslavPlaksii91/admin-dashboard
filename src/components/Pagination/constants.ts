export const ROWS_PER_PAGE = 8;
export const ROWS_PER_PAGE_OPTIONS = [8, 12, 16, 20];
