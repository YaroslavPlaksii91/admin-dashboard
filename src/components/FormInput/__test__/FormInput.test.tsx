import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import { createRegister } from '@utils/testMocks';

import { TOGGLE_BUTTON_TEST_ID } from '../constants';
import { FormInput } from '../FormInput';

const register = createRegister('name');

jest.mock('react-hook-form', () => ({
  useForm: () => ({
    register,
  }),
}));

describe('FormInput component', () => {
  afterEach(cleanup);

  const textInputProps = {
    label: 'Label',
    placeholder: 'Placeholder',
    name: 'name',
    type: 'text',
    errors: {},
    isPassword: false,
    register,
  };

  const passwordInputProps = {
    ...textInputProps,
    type: 'password',
    isPassword: true,
  };

  it('should rendered text type input', () => {
    const { asFragment } = render(<FormInput {...textInputProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render password type input', () => {
    const { asFragment } = render(<FormInput {...passwordInputProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should be rendered correctly', () => {
    render(<FormInput {...textInputProps} />);

    const inputElement = screen.getByLabelText(textInputProps.label);
    const placeholderElement = screen.getByPlaceholderText(
      textInputProps.placeholder,
    );

    expect(inputElement).toBeInTheDocument();
    expect(placeholderElement).toBeInTheDocument();
    expect(inputElement).toHaveAttribute('type', textInputProps.type);
  });

  it('toggles password visibility when it is a password input', () => {
    render(<FormInput {...passwordInputProps} />);

    const inputElement = screen.getByLabelText(passwordInputProps.label);
    const toggleButton = screen.getByTestId(TOGGLE_BUTTON_TEST_ID);

    expect(inputElement).toHaveAttribute('type', passwordInputProps.type);

    fireEvent.click(toggleButton);

    expect(inputElement).toHaveAttribute('type', textInputProps.type);
  });

  it('does not toggle password visibility when it is not a password input', () => {
    render(<FormInput {...textInputProps} />);

    const inputElement = screen.getByLabelText(textInputProps.label);
    const toggleButton = screen.queryByTestId(TOGGLE_BUTTON_TEST_ID);

    expect(inputElement).not.toHaveAttribute('type', passwordInputProps.type);
    expect(toggleButton).toBeNull();
  });

  it('shows the value entered by the user', () => {
    const enteredValue = 'Text value';

    render(<FormInput {...textInputProps} />);

    const inputElement = screen.getByLabelText(textInputProps.label);

    fireEvent.change(inputElement, { target: { value: enteredValue } });

    expect(screen.getByDisplayValue(enteredValue)).toBeInTheDocument();
  });

  it('shows default value if it is', () => {
    const defaultValue = 'Default value';

    render(<FormInput {...{ ...textInputProps, defaultValue }} />);

    const value = screen.getByDisplayValue(defaultValue);

    expect(value).toBeInTheDocument();
  });
});
