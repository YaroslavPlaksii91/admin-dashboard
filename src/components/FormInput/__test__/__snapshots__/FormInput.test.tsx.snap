// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput component should be rendered 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    mock="[object Object]"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r0:"
      id=":r0:-label"
    >
      Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        placeholder="Placeholder"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FormInput component should render password type 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    mock="[object Object]"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r1:"
      id=":r1:-label"
    >
      Password
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r1:"
        placeholder="Placeholder"
        type="password"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="toggle password visibility"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
          data-testid="toggleButton"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="VisibilityIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Password
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
